design/oauth2-device-authorization-grant.md

```diff
@@ -0,0 +1,144 @@
+
+## Endpoint
+
+### Device Authorization Endpoint
+
+In the spec, Device Authorization Endpoint is introduced for authorization request from the device which is an OAuth client. Keycloak has common AuthorizationEndpointBase class which provides common logic like checking realm, SSL and so on. We implement this new endpoint using the common base class. The URL is defined as `/realms/{realmName}/openid-connect/device/auth` in current implementation proposal.
+
+### Verification Endpoint
+
+Keycloak must provide a new endpoint for the verification process of the user code which is returned by device authorization request. Also, it's good UX that providing a shorter verification URI because the end-user need to enter the URI manually into their browser if the device has a limited display. So, it's defined as `/realms/{realmName}/device` in current implementation proposal.
+
+This endpoint provides user code verification and user authentication, which we will discuss in User Interaction section.
+
+### Token Endpoint
+
+Token request for the spec is represented with new grant type of `urn:ietf:params:oauth:grant-type:device_code`. We add the process of this new grant type into an existing TokenEndpoint class for OAuth 2.0/OIDC.
+
+
+## User Interaction when verifying a user code
+
+The spec says that the details of the user interaction when processing the verification are up to the authorization server in [3.3. User Interaction](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-3.3). There are two methods for the process which are imeplemented by major IdPs. *Login First* or *User Code Verification First*.
+
+### Login First
+
+In this case, the end-user need to login first before entering a user code in the user interaction. It means the authorization server must provide a shared authentication flow because the server can't detect the client ID yet which is bind to the user code. However, the risk getting user code brute force attack is low as attackers must log in first.
+
+### User Code Verification First
+
+In this case, the end-user need to enter a user code first in the user interaction. It means the authorization server can provide different authentication flow per client because the server can detect the client ID from the user code. However, the risk getting user code brute force attack is higher than Login First as attackers can try to enter user codes freely.
+
+### Which user interaction does keycloak should support?
+
+User Code Verification First is adopted by major IdPs like Google, Microsoft, and Salesforce. So it seems to be a common practice. Also, it can provide a more flexible authentication flow per device client. Keycloak already provides a feature that a client can use specific browser flow by using the Authentication Flow Overrides option. Therefore, it's suitable for supporting User Code Verification First.
+
+## Consent
```

Comment:
The fact that the user types the `user_code` when doing `verification_uri` does not imply the user is also consenting access to a device. The device can ask for scopes when making authorization requests and the user should be aware of the scopes he will grant to the device.

In both cases, the consent page should be displayed accordingly with the client configuration in regards to whether or not consent should be displayed. Just like it works today.

Reply:
The way I see the user verification flow is as follows:

* The user provides the code (directly or indirectly when doing `verification_uri_`complete`) at `/realms/{realms}/device`
* Once the code is submitted by the user, the `/realms/{realms}/device` will redirect the user to the login page. At this point, the device endpoint will set some state in a way that after the user is authenticated, we can mark the code as approved.

My point here is that the device endpoint is basically reusing and relying on the regular authentication flow configured to the realm with a caveat that we need some implicit execution during the flow that is capable of checking the user code being approced to allow the device to obtain an access token. 

Another approach is supporting a callback URI to the device endpoint. In this case, once the user authenticates, Keycloak will redirect the user back to the device endpoint in order to complete the user code approval.

Reply:
I believe we are saying the same thing, but neither of us are saying it 100% clearly. This is the flow I have in mind:

1. If verification_uri then user is asked to enter user code. If verification_uri_complete the user is asked to confirm if they want to link/authenticate the device.
2. User is asked to login if not already logged-in (through the regular browser flow)
3. User is asked to consent to scopes if client requires consent (through the regular browser flow)
4. Code is marked as approved

Reply:
Sure Stian, in fact, the flow is part of the standard :)

The point I'm trying to make is more related to the implementation details.  Although I didn't look at the current implementation yet.

Reply:
One of the reasons for displaying the consent page always is to prevent remote phishing, as mentioned below in the spec [5.4](https://tools.ietf.org/html/draft-ietf-oauth-device-flow-15#section-5.4).

```
5.4.  Remote Phishing

   It is possible for the device flow to be initiated on a device in an
   attacker's possession.  For example, an attacker might send an email
   instructing the target user to visit the verification URL and enter
   the user code.  To mitigate such an attack, it is RECOMMENDED to
   inform the user that they are authorizing a device during the user
   interaction step (see Section 3.3), and to confirm that the device is
   in their possession.  The authorization server SHOULD display
   information about the device so that the person can notice if a
   software client was attempting to impersonating a hardware device.
```

But, I agree to use the existing `Consent Required` option. However, in that case, I would like to propose adding the `Display Consent Always` sub-option which is configurable when `Consent Required` is true. In the current implementation, once the user agrees, it will be skipped in the subsequent authentication.

Reply:
I'd say that is not needed as there is a separate consent to authorize the device already present, which is either user entering the user code manually or confirming the user code. When entering the code manually the user knows the device as the code is coming from the device. When reviewing the code we know the client so can display info about the client alongside the user code.

Reply:
@stianst 

> When reviewing the code we know the client so can display info about the client alongside the user code.

You mean the end-user can see the screen for reviewing after entering user code even if verification_uri is used?  The flow is as follows. If so, I agree.

1. User is asked to enter user code. then enter user code.
2. If the user code is valid, user is asked to continue with information including the device (client information).
3. If user click continue button,  user is asked to login if not already logged-in (through the regular browser flow).
4. User is asked to consent to scopes if client requires consent (through the regular browser flow).
5. Device code is marked as approved.

Also, to implement review screen in step 2 above, should we add `User Code Review Screen Text` option in client config page?

Reply:
With verification_uri the user already knows the device as the user directly copies the code from the device, so I don't think there is a need for a separate review page here. By entering the code manually the user implicitly authorizes the device.

The authorization page for the device should only be displayed for verification_uri_complete as there is no manual step required for that one.

Reply:
@stianst 

>With verification_uri the user already knows the device as the user directly copies the code from the device, so I don't think there is a need for a separate review page here.

In the spec, it is also assumed that the device might send the verification_uri (not verification_uri_complete) to the user by email, etc. instead of manually inputting the URI. In this case, the user clicks the URI in the email. Due to the risk of phishing emails, it would be dangerous for the user to consider it as agreed by clicking on the URI. That's why the spec recommends informing the user that they are authorizing a device during the user interaction.

Reply:
Sure, but the user would click the link verification_uri in the email, which would display a message saying that you are authorizing a device and ask to enter the user code before continuing. There's no need for yet another step.

Reply:
It seems you are basically moving the "user's device verification step" from Keycloak to the e-mail that the device may send to the user. If so, I understand your point.

However, I still think that we should enforce that step in Keycloak. The reason being that a malicious device/attacker can craft an email and send to the user. By forcing the verification step in Keycloak we are forcing users to check if the device is really in their possession and trustworthy.

Reply:
> Sure, but the user would click the link verification_uri in the email, which would display a message saying that you are authorizing a device and ask to enter the user code before continuing. There's no need for yet another step.

The `verification_uri` is a common static link for all users in the realm, like `/realms/{realmName}/device`. So the user can't notice the device they try to authorize until entering a valid user code.
As @pedroigor mentioned above, it makes sense to add one more step.

Reply:
I disagree that the extra step is required. When opening verification_uri the user needs to retrieve the user code from somewhere. Users should obviously only retrieve this code from a device they trust and not from random places. By copying the user code from the display of a device a user explicitly trusts the device as such it is not needed to ask the user yet again to trust that device.

Further, in most cases we are talking about public clients/devices. As such anyone can just use the client-id for a "trusted device". So displaying the display name after a user has retrieved the user code from an malicious source would only give the illusion to the user that they are linking to a trusted device.